connectors/protocol/connectors.py

```diff
@@ -621,6 +621,10 @@ async def sync_done(self, job):
             "error": job_error,
         }
 
+        # only update sync cursor after a successful sync job
+        if job_status == JobStatus.COMPLETED:
```

Comment:
I have 2 comments here: 

1. Would it be possible for a sync job that was not completed with success to report the current progress? So the next job can pick it up? 
2. I have a feeling that `sync_done` should not decide if the cursor needs to be saved or not. This is just a "reporting" method so I would not add any logic into it.

Reply:
>Would it be possible for a sync job that was not completed with success to report the current progress? So the next job can pick it up?

I considered this option when drafting the implementation doc, but I decided to only update the cursor after a successful job.

E.g. A sync job fetched 100 docs and pushed them into the queue, then encountered an error. Let's see 50 out of the 100 docs are indexed into ES. If we update the `sync_cursor` when failing the sync job, it will miss the 50 docs in the queue when running the next sync job.

So it's much safer to update `sync_cursor` only after a successful sync job. We can iterate further in the later release.

Reply:
> I have a feeling that `sync_done` should not decide if the cursor needs to be saved or not. This is just a "reporting" method so I would not add any logic into it.

IMO, `sync_done` is to update the connector when a sync job is done (either successfully or not), so it contains logic to handle different scenarios, e.g. 1) make sure `last_sync_error` is set if the job status is `error`, 2) only update `last_indexed_document_count` and `last_deleted_document_count` when the sync job is terminated. So I feel it makes sense to have this check here.

Reply:
> I considered this option when drafting the implementation doc, but I decided to only update the cursor after a successful job.
E.g. A sync job fetched 100 docs and pushed them into the queue, then encountered an error. Let's see 50 out of the 100 docs are indexed into ES. If we update the sync_cursor when failing the sync job, it will miss the 50 docs in the queue when running the next sync job.
So it's much safer to update sync_cursor only after a successful sync job. We can iterate further in the later release.

Get it. It makes sense to me for this iteration. What do you do when the incremental sync failed? Can we just restart it without changing the cursor?

Reply:
>What do you do when the incremental sync failed? Can we just restart it without changing the cursor?

It just fails. `error` is a terminal status. Customers have to trigger/schedule another incremental sync, which will take the cursor from the connector, i.e. the same cursor as the failed incremental sync.

Reply:
> IMO, sync_done is to update the connector when a sync job is done (either successfully or not), so it contains logic to handle different scenarios, e.g. 1) make sure last_sync_error is set if the job status is error, 2) only update last_indexed_document_count and last_deleted_document_count when the sync job is terminated. So I feel it makes sense to have this check here.

`sync_done` is already overloaded with business logic. Adding more if/else conditions to it will make it even more complex. Let's imagine we want to update cursors when sync is failed then you will need to add another if/else to the `error` or `sync_done` methods and so on. In my ideal world it should look like this: 

```
def sync_done(args)
  doc = update_status(doc, args)
  doc = update_error(doc, args)
  doc = update_cursor(doc, args)
  
  store(doc)
```

So I would at least incapsulate updating cursor logic even if you want to call inside `sync_done`. 

_It's not a blocker as we need to move forward._